libpod/image/image.go

```diff
@@ -131,41 +142,72 @@ func (ir *Runtime) NewFromLocal(name string) (*Image, error) {
 	return &image, nil
 }
 
+// DefaultPullConfig returns a default new image config for runtime.New()
+func (ir *Runtime) DefaultPullConfig(name string) *PullConfig {
+	imgCtx := new(PullConfig)
+	imgCtx.Name = name
+	imgCtx.SignaturePolicyPath = ""
+	imgCtx.Authfile = ""
+	imgCtx.Writer = nil
+	imgCtx.DockerOptions = nil
+	imgCtx.SigningOptions = SigningOptions{}
+	imgCtx.Label = nil
+	return imgCtx
+}
+
 // New creates a new image object where the image could be local
 // or remote
-func (ir *Runtime) New(ctx context.Context, name, signaturePolicyPath, authfile string, writer io.Writer, dockeroptions *DockerRegistryOptions, signingoptions SigningOptions, forcePull bool, label *string) (*Image, error) {
+func (ir *Runtime) New(ctx context.Context, config *PullConfig, forcePull, update bool) (*Image, error) {
 	span, _ := opentracing.StartSpanFromContext(ctx, "newImage")
 	span.SetTag("type", "runtime")
 	defer span.Finish()
 
 	// We don't know if the image is local or not ... check local first
 	newImage := Image{
-		InputName:    name,
+		InputName:    config.Name,
 		Local:        false,
 		imageruntime: ir,
 	}
+
+	if config.SignaturePolicyPath == "" {
+		config.SignaturePolicyPath = ir.SignaturePolicyPath
+	}
+
+	var imageName []string
+	var err error
+
 	if !forcePull {
 		localImage, err := newImage.getLocalImage()
 		if err == nil {
-			newImage.Local = true
-			newImage.image = localImage
-			return &newImage, nil
+			upToDate := true
+			if update {
+				// check if it needs updating, and if so update it
+				imageName, upToDate, err = ir.pullImageIfOutOfDate(ctx, localImage, config)
```

Comment:
Looking at the various, not really documented and somewhat scary, things `getLocalImage` does, it’s fairly likely that it will find some other image than the remote image `pullImageIfOutOfDate` does.

Even if `getLocalImage` were quite trivial and obviously correct, it still does not make sense to me to have _one_ local image matched against _many_ remote images (depending on the search path); that could only work if the search path contained mirrors with equivalent content, which is not the case.

`pullGoal` entries already contain a precise storage reference; so, I think `pullImageIfOutOfDate` can use _that_ reference to see if a local image exists, and if it does, obtain it in there inside that loop, and _then_ check whether it needs updating, and return the local image up into this function.